### PR TITLE
Implementation of data exchange mechanisms between robot and supervisor using Serial Webots drivers 

### PR DESCRIPTION
**Pull request by akrambzeo**
Main functions:

Initialization of the serial communication:

Initialize the serial driver in the robot and in the supervisor.

Sensor data transmission:

Sensor data is sent from the robot to the supervisor, and the supervisor receives the data.

Motor control:

The Supervisor sends the specified motor control data to the robot.
The robot receives the motor control data and updates the motor speeds accordingly.

Reset Logic for Loss of Line Tracking:

Monitor sensor data for loss of line tracking.
Initialize the robot's position and orientation after a fixed number of steps without line detection.
Reset the robot to the standby state without recalibrating the motors and sensors.

Training or driving mode:

The mode (training or driving) is selected by pressing a button.
If no button is pressed within a specified time, the robot will default to the training mode.
Training or driving will not begin until the start/stop line is crossed.

### CODE REVIEW

File: webots/controllers/Supervisor/SerialMuxProt.py
Comment (gabryelreyes) [inline webots/controllers/Supervisor/SerialMuxProt.py:1]: This file should not be copied from its original library. Please check the README of the library to perform the correct installation
https://github.com/gabryelreyes/SerialMuxProt

File: platformio.ini [hunk 1]
```diff
@@ -16,13 +16,14 @@
...
 ;default_envs = ConvoyLeaderTarget
-default_envs = ConvoyLeaderSim
+;default_envs = ConvoyLeaderSim
 ;default_envs = LineFollowerTarget
 ;default_envs = LineFollowerSim
 ;default_envs = RemoteControlTarget
 ;default_envs = RemoteControlSim
 ;default_envs = SensorFusionTarget
 ;default_envs = SensorFusionSim
+;default_envs = ReinforcementLearning
```
Comment (gabryelreyes) [inline platformio.ini:26]: Added the environment but still commented it out?

File: platformio.ini [hunk 2]
```diff
@@ -388,7 +442,7 @@ lib_deps =
 lib_ignore =
     APPCalib
     APPConvoyLeader
-    APPLineFollower
+    APPReinforcementLearning
```
Comment (gabryelreyes) [inline platformio.ini:445]: Should not remove APPLineFollower

File: webots/controllers/Supervisor/Supervisor.py
Comment (gabryelreyes) [inline webots/controllers/Supervisor/Supervisor.py:1]: Do not modify the same supervisor used by other applications. Create your own supervisor in a separate folder and revert the changes to this one.

File: webots/protos/Supervisor.proto
Comment (gabryelreyes) [inline webots/protos/Supervisor.proto:1]: See the comment on the supervisor controller

File: platformio.ini [hunk 3]
```diff
@@ -282,6 +286,54 @@ lib_ignore =
...
+lib_deps =
+    ${hal:Target.lib_deps}
+    HALLineFollowerTarget
+lib_ignore =
+    ${hal:Target.lib_ignore}
+extra_scripts =
+    ${hal:Target.extra_scripts}
+
+; *****************************************************************************
+; Line follower application specific HAL for simulation
```
Comment (gabryelreyes) [inline platformio.ini:323]: Wrong comment. Check everywhere else

**Review by gabryelreyes — CHANGES_REQUESTED** [review]
Did not review content but structure and integrity of other apps.

- All files under the path ` lib/APPReinforcementLearning/`  and ` lib/HALReinforcementLearning/`should be contained in a ` src` folder. Check the structure in other apps.

File: webots/controllers/RL_Supervisor/RL_Supervisor.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+supervisor = Supervisor()
+
+# Get the time step of the current world.
+timestep = int(supervisor.getBasicTimeStep())
+
+# Get the Supervisor Receiver Device
+supervisor_com_rx = supervisor.getDevice(SUPERVISOR_RX_NAME)
+
+# Get the Supervisor Emitter device 
+supervisor_com_Tx = supervisor.getDevice(SUPERVISOR_TX_NAME)
```
Comment (hoeftjch) [inline webots/controllers/RL_Supervisor/RL_Supervisor.py:29]: Adhere PEP 8 style guide.

Imports:
Imports should be grouped into standard library imports, related third-party imports, and local application/library-specific imports. Each group should be separated by a blank line.

Constants:
Constants should be defined in uppercase with words separated by underscores.

Variable Naming:
Variable names should be in snake_case.

Global Variables:
Avoid using global variables. If necessary, encapsulate them within a class or pass them as parameters.
String Formatting:

Indentation:
Ensure consistent indentation. Use 4 spaces per indentation level.

Comments:
Comments should be complete sentences and start with a capital letter.

File: webots/controllers/RL_Supervisor/RL_Supervisor.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+    driving_mode = payload[0]
+    if driving_mode:
+        print("Driving Mode Selected")
+    else:
+        print("Train Mode Selected")
+
+
+m_serialMuxProtChannelIdSPEED_SET = smp_server.create_channel("SPEED_SET", SPEED_SET_DLC)
+m_serialMuxProtChannelIdSTATUS = smp_server.create_channel("STATUS", STATUS_DLC)
+smp_server.subscribe_to_channel("STATUS", callback_Status)
```
Comment (hoeftjch) [inline webots/controllers/RL_Supervisor/RL_Supervisor.py:116]: use constants for channel names

**Review by hoeftjch — CHANGES_REQUESTED** [review]
Review of C++ part haltet for restructuring and fixing previous existing review comments

File: lib/APPReinforcementLearning/App.cpp
Comment (gabryelreyes) [inline lib/APPReinforcementLearning/App.cpp:1]: File is not in src folder

File: lib/HALReinforcementLearningSim/Board.cpp
Comment (gabryelreyes) [inline lib/HALReinforcementLearningSim/Board.cpp:1]: File is not in src folder

File: lib/HALReinforcementLearningSim/Board.h
Comment (gabryelreyes) [inline lib/HALReinforcementLearningSim/Board.h:1]: File is not in src folder

File: platformio.ini [hunk 6]
```diff
@@ -195,6 +197,7 @@ lib_ignore =
     APPConvoyFollower
     APPLineFollower
     APPRemoteControl
+    APPReinforcementLearning
```
Comment (gabryelreyes) [inline platformio.ini:200]: Please keep this list in alphabetical order. Fix everywhere

File: platformio.ini [hunk 7]
```diff
@@ -343,6 +349,40 @@ lib_ignore =
...
+; Reinforcement Learning for Line follower application
+; *****************************************************************************
+[app:ReinforcementLearning]
+build_flags =
+    ${common.build_flags}
+    -DLOG_DEBUG_ENABLE
+lib_deps =
+    APPReinforcementLearning
+    Service
+    gabryelreyes/SerialMuxProt @ ~2.2.0
```
[the remark this answers is not in the document]
Comment (gabryelreyes) [reply platformio.ini:363]: This tilde before the version means that the patches are accepted. this means the version 2.2.2 is already used

File: platformio.ini [hunk 8]
```diff
@@ -343,6 +349,40 @@ lib_ignore =
...
+    APPConvoyLeader
+    APPRemoteControl
+    APPSensorFusion
+    APPLineFollower
+    APPTest
+
+; *****************************************************************************
+; Reinforcement Learning for Line follower application specific HAL for simulation
+; *****************************************************************************
+[hal_app:ReinforcementLearningSim]
```
Comment (gabryelreyes) [inline platformio.ini:375]: Please maintain the correct order of the file. the `hal_app:` comes before the `app:`

File: webots/protos/NTrack.png
Comment (gabryelreyes) [inline webots/protos/NTrack.png:1]: What does the `N` in the name stands for? Use more descriptive names please

**Review by gabryelreyes — CHANGES_REQUESTED** [review]

File: webots/controllers/RL_Supervisor/RL_Supervisor.py (hunk 4 above)
Comment (gabryelreyes) [reply webots/controllers/RL_Supervisor/RL_Supervisor.py:29]: Ensure you have PyLint Extension installed to provide support while adhering to the PEP8 style guide.
Style will be also checked by the CI (#136).

File: lib/APPReinforcementLearning/src/App.cpp [hunk 9]
```diff
@@ -1,46 +1,29 @@
-/* MIT License
- *
```
Comment (gabryelreyes) [inline lib/APPReinforcementLearning/src/App.cpp:2]: Why removed the license? Please re-add it everywhere where missing

**Review by gabryelreyes — CHANGES_REQUESTED** [review]

File: lib/APPReinforcementLearning/src/App.h [hunk 10]
```diff
@@ -75,14 +86,21 @@ class App
 
     /**
      * Setup the application.
-     */
+    */
```
Comment (gabryelreyes) [inline lib/APPReinforcementLearning/src/App.h:89]: Please do not change the formatting of the comments. Fix everywhere

File: lib/APPReinforcementLearning/src/App.h [hunk 11]
```diff
@@ -97,22 +115,49 @@ class App
...
+    bool setupSerialMuxProt();
+    
+    /**
+     * Send line sensors data via SerialMuxProt.
+    */
+    void sendLineSensorsData() const;
+
+    /* Not allowed. */
+    App(const App& app);            /**< Copy construction of an instance. */
+    App& operator=(const App& app); /**< Assignment of an instance. */
```
Comment (gabryelreyes) [inline lib/APPReinforcementLearning/src/App.h:160]: Why modified the comment of these methods? This is old, please change back to the new comment

File: lib/APPReinforcementLearning/src/ErrorState.h [hunk 12]
```diff
@@ -122,23 +122,9 @@ class ErrorState : public IState
     {
     }
 
-    /**
```
Comment (gabryelreyes) [inline lib/APPReinforcementLearning/src/ErrorState.h:125]: Do not change the comments of these functions. Fix everywhere

File: lib/APPReinforcementLearning/src/LineSensorsCalibrationState.cpp [hunk 13]
```diff
@@ -40,6 +40,7 @@
 #include <Util.h>
 #include "ReadyState.h"
 #include "ErrorState.h"
+#include "DrivingState.h"
```
Comment (gabryelreyes) [inline lib/APPReinforcementLearning/src/LineSensorsCalibrationState.cpp:43]: Why this here?

File: lib/APPReinforcementLearning/src/LineSensorsCalibrationState.h [hunk 14]
```diff
@@ -131,23 +131,9 @@ class LineSensorsCalibrationState : public IState
     {
     }
 
-    /**
-     * Copy construction of an instance.
```
Comment (gabryelreyes) [inline lib/APPReinforcementLearning/src/LineSensorsCalibrationState.h:135]: See comment in ErrorState.h

File: lib/APPReinforcementLearning/src/MotorSpeedCalibrationState.h [hunk 15]
```diff
@@ -137,23 +137,9 @@ class MotorSpeedCalibrationState : public IState
     {
     }
 
-    /**
-     * Copy construction of an instance.
-     * Not allowed.
-     *
-     * @param[in] state Source instance.
```
Comment (gabryelreyes) [inline lib/APPReinforcementLearning/src/MotorSpeedCalibrationState.h:144]: See comment in ErrorState.h

File: lib/APPReinforcementLearning/src/ReadyState.cpp [hunk 16]
```diff
@@ -151,18 +148,40 @@ void ReadyState::setLapTime(uint32_t lapTime)
...
 
-/******************************************************************************
- * Protected Methods
- *****************************************************************************/
+uint8_t ReadyState::setSelectedMode()
+{
+    return (m_mode);
+}
 
-/******************************************************************************
```
Comment (gabryelreyes) [inline lib/APPReinforcementLearning/src/ReadyState.cpp:158]: Why deleted the section separators? Revert

File: lib/APPReinforcementLearning/src/ReadyState.h [hunk 17]
```diff
@@ -96,17 +98,86 @@ class ReadyState : public IState
...
+    static const uint8_t SENSOR_ID_MOST_LEFT;
+
+    /**
+     * ID of middle sensor.
+    */
+    static const uint8_t SENSOR_ID_MIDDLE;
+
+    /**
+     * ID of most right sensor.
+    */
```
Comment (gabryelreyes) [inline lib/APPReinforcementLearning/src/ReadyState.h:134]: Fix comment formatting. Fix everywhere

File: lib/APPReinforcementLearning/src/ReadyState.h [hunk 18]
```diff
@@ -96,17 +98,86 @@ class ReadyState : public IState
...
+    /** Timeout timer for the selected mode. */
+    SimpleTimer m_modeTimeoutTimer; 
+
+    /** The system state machine. */
+    StateMachine m_systemStateMachine;  
+
+    /**< Is set (true), if a lap time is available. */
+    bool m_isLapTimeAvailable; 
+
+    /**< Lap time in ms of the last successful driven round. */
```
Comment (gabryelreyes) [inline lib/APPReinforcementLearning/src/ReadyState.h:156]: Wrong comment style

File: lib/APPReinforcementLearning/src/ReadyState.h [hunk 19]
```diff
@@ -96,17 +98,86 @@ class ReadyState : public IState
...
+    */
+    bool LastLineStatus ;
+
+    /** Timeout timer for the selected mode. */
+    SimpleTimer m_modeTimeoutTimer; 
+
+    /** The system state machine. */
+    StateMachine m_systemStateMachine;  
+
+    /**< Is set (true), if a lap time is available. */
```
Comment (gabryelreyes) [inline lib/APPReinforcementLearning/src/ReadyState.h:153]: Wrong comment style

File: lib/APPReinforcementLearning/src/ReadyState.h [hunk 20]
```diff
@@ -117,28 +188,30 @@ class ReadyState : public IState
...
     }
 
-    /**
-     * Copy construction of an instance.
-     * Not allowed.
-     *
-     * @param[in] state Source instance.
-     */
-    ReadyState(const ReadyState& state);
+    /* Not allowed. */
```
Comment (gabryelreyes) [inline lib/APPReinforcementLearning/src/ReadyState.h:191]: See comment in ErrorState.h

File: lib/APPReinforcementLearning/src/SerialMuxchannels.h [hunk 21]
```diff
@@ -0,0 +1,154 @@
...
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+/*******************************************************************************
+    DESCRIPTION
+*******************************************************************************/
+/**
+ *  @brief  Channel structure definition for the SerialMuxProt.
+ *  @author Gabryel Reyes <gabryelrdiaz@gmail.com>
```
Comment (gabryelreyes) [inline lib/APPReinforcementLearning/src/SerialMuxchannels.h:29]: You are the author

File: lib/APPReinforcementLearning/src/StartupState.h [hunk 22]
```diff
@@ -135,23 +129,9 @@ class StartupState : public IState
...
-    /**
-     * Copy construction of an instance.
-     * Not allowed.
-     *
-     * @param[in] state Source instance.
-     */
-    StartupState(const StartupState& state);
-
-    /**
-     * Assignment of an instance.
```
Comment (gabryelreyes) [inline lib/APPReinforcementLearning/src/StartupState.h:147]: See comment in ErrorState.h

File: lib/HALReinforcementLearningSim/src/Board.h [hunk 23]
```diff
@@ -239,7 +240,7 @@ class Board : public IBoard
     {
         m_buzzer.process();
     }
-
+
```
Comment (gabryelreyes) [inline lib/HALReinforcementLearningSim/src/Board.h:243]: Remove extra space

File: lib/HALReinforcementLearningSim/src/Board.h [hunk 24]
```diff
@@ -287,6 +288,11 @@ class Board : public IBoard
     /** Settings */
     Settings m_settings;
 
+    /** Simulation serial driver */
+    WebotsSerialDrv m_serialDrv;
+    
+
```
Comment (gabryelreyes) [inline lib/HALReinforcementLearningSim/src/Board.h:294]: Remove extra space

File: lib/HALReinforcementLearningSim/src/Board.h [hunk 25]
```diff
@@ -299,13 +305,12 @@ class Board : public IBoard
      */
     Board();
 
-    /**
+       /**
```
Comment (gabryelreyes) [inline lib/HALReinforcementLearningSim/src/Board.h:308]: Fix comment formatting

File: lib/HALReinforcementLearningSim/src/Board.h [hunk 26]
```diff
@@ -299,13 +305,12 @@ class Board : public IBoard
...
      * Destroys the concrete board.
      */
-    ~Board()
```
Comment (gabryelreyes) [inline lib/HALReinforcementLearningSim/src/Board.h:305]: Fix formatting

**Review by gabryelreyes — CHANGES_REQUESTED** [review]

**Comment by gabryelreyes** [comment]
@akrambzeo Please fix all CI errors

File: lib/APPReinforcementLearning/src/DrivingState.h [hunk 27]
```diff
@@ -26,7 +26,7 @@
 *******************************************************************************/
 /**
  * @brief  Driving state
- * @author Andreas Merkle <web@blue-andi.de>
+ * @author @author Akram Bziouech
```
Comment (gabryelreyes) [inline lib/APPReinforcementLearning/src/DrivingState.h:29]: Extra author tag

File: lib/APPReinforcementLearning/src/ErrorState.cpp [hunk 28]
```diff
@@ -60,21 +59,14 @@
...
 /******************************************************************************
  * Public Methods
  *****************************************************************************/
 
 void ErrorState::entry()
 {
-    IBoard&   board   = Board::getInstance();
-    IDisplay& display = board.getDisplay();
-
-    /* Stop the motors in any case! */
```
Comment (gabryelreyes) [inline lib/APPReinforcementLearning/src/ErrorState.cpp:77]: Unnecessary change

File: lib/APPReinforcementLearning/src/SerialMuxchannels.h [hunk 29]
```diff
@@ -0,0 +1,152 @@
...
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+/*******************************************************************************
+    DESCRIPTION
+*******************************************************************************/
+/**
+ *  @brief  Channel structure definition for the SerialMuxProt.
+ *  @author @author Akram Bziouech
```
Comment (gabryelreyes) [inline lib/APPReinforcementLearning/src/SerialMuxchannels.h:29]: Extra author tag

**Review by gabryelreyes — CHANGES_REQUESTED** [review]

**Comment by gabryelreyes** [comment]
Please run the clang formatter from VSCode on all your files: `alt + shift + D` . This should fix some of the formatting issues.

**Comment by gabryelreyes** [comment]
Please add your application to the README.md in the "The Applications" section

File: webots/controllers/RL_Supervisor/Serial_webots.py [hunk 30]
```diff
@@ -1,13 +1,41 @@
+""" Implementation of a Serial Webots for Serial Communication """
+
+# MIT License
+#
+# Copyright (c) 2023 - 2024 Gabryel Reyes <gabryelrdiaz@gmail.com>
```
Comment (gabryelreyes) [inline webots/controllers/RL_Supervisor/Serial_webots.py]: Fix the owner of the repository license

File: README.md [hunk 31]
```diff
@@ -13,7 +13,7 @@ Several kind of exclusive applications are available:
 * Line Follower - Just a line follower, using a PID controller.
 * Remote Control - The robot is remote controlled by e.g. the [DroidControlShip](https://github.com/BlueAndi/DroidControlShip) in a convoy follower role.
 * Sensor Fusion - The robot provides odometry and inertial data to the [DroidControlShip](https://github.com/BlueAndi/DroidControlShip), which calculates the sensor fusion based location information.
-
+* Line Follower with Reinforcement Learning - A line follower with reinforcement learning uses a learning agent that uses rewards and punishments to steer optimally in order to follow a line autonomously.
```
Comment (gabryelreyes) [inline README.md:16]: Missing new line

File: README.md [hunk 32]
```diff
@@ -158,6 +158,8 @@ Example for the **LineFollowerTarget** application:
 | LineFollowerSimple | Just a simple line follower, using a PID controller. | Yes | No | ./webots/worlds/ETrack.wbt  ./webots/worlds/LargeTrack.wbt  ./webots/worlds/LineFollowerTrack.wbt |
 | RemoteControl | The robot is remote controlled by e.g. the [DroidControlShip](https://github.com/BlueAndi/DroidControlShip) in a convoy follower role. | No | Yes | ./webots/world/zumo_with_com_system/* |
 | SensorFusion | The robot provides odometry and inertial data to the [DroidControlShip](https://github.com/BlueAndi/DroidControlShip), which calculates the sensor fusion based location information. | No | Yes | ./webots/worlds/zumo_with_com_system/LineFollowerTrack.wbt |
+| ReinforcementLearning | A line follower with reinforcement learning uses a learning agent that uses rewards and punishments to steer optimally in order to follow a line autonomously. 
+| Yes | No | ./webots/worlds/RL_LineFollower.wbt |
```
Comment (gabryelreyes) [inline README.md:162]: Wrong table formatting. 2 Rows added instead of one

File: webots/controllers/RL_Supervisor/serial_webots.py [hunk 33]
```diff
@@ -1,13 +1,41 @@
+""" Implementation of a Serial Webots for Serial Communication """
+
+# MIT License
+#
+# Copyright (c) 2023 - 2024 Gabryel Reyes <gabryelrdiaz@gmail.com>
```
Comment (gabryelreyes) [inline webots/controllers/RL_Supervisor/serial_webots.py:5]: Fix the owner of the license

File: webots/controllers/RL_Supervisor/rl_supervisor.py [hunk 34]
```diff
@@ -0,0 +1,214 @@
+"""
+Supervisor controller.
+
+Details: https://github.com/cyberbotics/webots/blob/master/docs/guide/supervisor-programming.md
+"""
+
```
Comment (gabryelreyes) [inline webots/controllers/RL_Supervisor/rl_supervisor.py:6]: Add the license to this file and the sections documentation

**Review by gabryelreyes — CHANGES_REQUESTED** [review]

File: lib/APPReinforcementLearning/src/DrivingState.cpp [hunk 35]
```diff
@@ -26,21 +26,17 @@
...
  */
 
 /******************************************************************************
  * Includes
  *****************************************************************************/
 #include "DrivingState.h"
-#include <Board.h>
-#include <Sound.h>
-#include <DifferentialDrive.h>
 #include <StateMachine.h>
```
Comment (gabryelreyes) [inline lib/APPReinforcementLearning/src/DrivingState.cpp:36]: Unnecessary change of the order of the includes

File: lib/APPReinforcementLearning/src/DrivingState.h [hunk 36]
```diff
@@ -91,114 +90,34 @@ class DrivingState : public IState
...
-        TRACK_STATUS_TRACK_LOST_BY_GAP,       /**< Track lost by gap. */
-        TRACK_STATUS_TRACK_LOST_BY_MANOEUVRE, /**< Track lost by driving manoeuvre. */
-        TRACK_STATUS_FINISHED                 /**< Robot found the end line or a error happened. */
-    };
+    bool isAbortRequired();
 
+protected:
+private:
     /** Observation duration in ms. This is the max. time within the robot must be finished its drive. */
     static const uint32_t OBSERVATION_DURATION = 3000000;
```
Comment (gabryelreyes) [inline lib/APPReinforcementLearning/src/DrivingState.h:111]: How do you get this number? Can you decompose it in easier numbers to work with? (Minutes, for example)

File: lib/APPReinforcementLearning/src/ReadyState.cpp [hunk 37]
```diff
@@ -59,89 +59,87 @@
...
-
-        Util::intToStr(valueStr, sizeof(valueStr), position);
-        LOG_DEBUG_MSG(valueStr);
-
-        LOG_DEBUG_TAIL();
-
-        m_timer.restart();
+        m_mode = TRAINING_MODE;
+    }
+    else if (true == m_modeTimeoutTimer.isTimeout() && (m_mode == IDLE))
```
Comment (gabryelreyes) [inline lib/APPReinforcementLearning/src/ReadyState.cpp:117]: Why TRAINING_MODE after timeout?

File: lib/APPReinforcementLearning/src/ReadyState.cpp [hunk 38]
```diff
@@ -59,89 +59,87 @@
...
+        m_modeTimeoutTimer.restart();
     }
     else
     {
         /* Nothing to do. */
         ;
     }
+
+    /**Drive forward until START LINE is crossed */
+    DriveUntilStartLineisCrossed();
```
Comment (gabryelreyes) [inline lib/APPReinforcementLearning/src/ReadyState.cpp:129]: functions shall start with lowercase

File: lib/APPReinforcementLearning/src/StartupState.h [hunk 39]
```diff
@@ -115,7 +115,6 @@ class StartupState : public IState
     UserInfo    m_userInfoState; /**< Current user info state. */
     bool        m_isButtonAPressed; /**< Is the button A pressed (last time)? */
     bool        m_isButtonBPressed; /**< Is the button B pressed (last time)? */
-
```
Comment (gabryelreyes) [inline lib/APPReinforcementLearning/src/StartupState.h:118]: Add deleted line

File: webots/controllers/RL_Supervisor/rl_supervisor.py [hunk 40]
```diff
@@ -0,0 +1,237 @@
...
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+# OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+
```
Comment (gabryelreyes) [inline webots/controllers/RL_Supervisor/rl_supervisor.py:29]: Add Section separators such as in `serial_webots`

**Review by gabryelreyes — CHANGES_REQUESTED** [review]

File: lib/APPReinforcementLearning/src/DrivingState.cpp (hunk 35 above)
Comment (akrambzeo) [reply lib/APPReinforcementLearning/src/DrivingState.cpp:36]: i have removed the include sound because i don't need it should i reset the order as before or does it fit like this

File: lib/APPReinforcementLearning/src/DrivingState.h (hunk 36 above)
Comment (akrambzeo) [reply lib/APPReinforcementLearning/src/DrivingState.h:111]: I have  used the same number as for Linefollower Application !

File: lib/APPReinforcementLearning/src/ReadyState.cpp (hunk 37 above)
Comment (gabryelreyes) [reply lib/APPReinforcementLearning/src/ReadyState.cpp:117]: Please add a comment to understand why this happens

File: lib/APPReinforcementLearning/src/DrivingState.h (hunk 36 above)
Comment (gabryelreyes) [reply lib/APPReinforcementLearning/src/DrivingState.h:111]: Ok

File: lib/APPReinforcementLearning/src/ReadyState.cpp (hunk 37 above)
Comment (akrambzeo) [reply lib/APPReinforcementLearning/src/ReadyState.cpp:117]: if button a or button b is not pressed it should automatically when timer is timeout, the training mode should be set and we discussed this at the previous meeting

File: lib/APPReinforcementLearning/src/DrivingState.cpp (hunk 35 above)
Comment (gabryelreyes) [reply lib/APPReinforcementLearning/src/DrivingState.cpp:36]: Yes. you can remove the sound without changing the order

**Review by gabryelreyes — APPROVED** [review]
Looks good to me. @hoeftjch should make the final lreview and approval before merging.
Edit: found some more stuff

File: lib/APPReinforcementLearning/src/App.cpp [hunk 41]
```diff
@@ -106,29 +110,76 @@ void App::loop()
...
+        (&DrivingState::getInstance() == m_systemStateMachine.getState()))
+    {
+        sendLineSensorsData();
 
-        m_controlInterval.restart();
+        m_sendLineSensorsDataInterval.restart();
+    }
+
+    /* Send Mode selected to The Supervisor. */
+    if (&ReadyState::getInstance() == m_systemStateMachine.getState() && (!m_modeSelectionSent))
```
Comment (gabryelreyes) [inline lib/APPReinforcementLearning/src/App.cpp:142]: Please use explicit comparison.
```suggestion
    if (&ReadyState::getInstance() == m_systemStateMachine.getState() && (false == m_modeSelectionSent))
```

File: lib/APPReinforcementLearning/src/App.cpp [hunk 42]
```diff
@@ -106,29 +110,76 @@ void App::loop()
...
+    {
+        uint8_t mode_options = ReadyState::getInstance().setSelectedMode();
+
+        if (mode_options > 0)
```
Comment (gabryelreyes) [inline lib/APPReinforcementLearning/src/App.cpp:146]: Use Yoda-style.
```suggestion
        if (0 < mode_options)
```

Comment (gabryelreyes) [reply lib/APPReinforcementLearning/src/App.cpp:146]: Fix everywhere

File: lib/APPReinforcementLearning/src/App.cpp [hunk 43]
```diff
@@ -106,29 +110,76 @@ void App::loop()
...
+        {
+            SMPChannelPayload::Mode payload =
+                (mode_options == 1) ? SMPChannelPayload::Mode::DRIVING_MODE : SMPChannelPayload::Mode::TRAINING_MODE;
```
Comment (gabryelreyes) [inline lib/APPReinforcementLearning/src/App.cpp:149]: Use Yoda-style.

File: lib/APPReinforcementLearning/src/App.cpp [hunk 44]
```diff
@@ -137,10 +188,89 @@ void App::loop()
...
 
+void App::sendLineSensorsData() const
+{
+    ILineSensors&   lineSensors      = Board::getInstance().getLineSensors();
+    uint8_t         maxLineSensors   = lineSensors.getNumLineSensors();
+    const uint16_t* lineSensorValues = lineSensors.getSensorValues();
+    uint8_t         lineSensorIdx    = 0U;
+    LineSensorData  payload;
+
+    if (LINE_SENSOR_CHANNEL_DLC == maxLineSensors * sizeof(uint16_t))
```
Comment (gabryelreyes) [inline lib/APPReinforcementLearning/src/App.cpp:199]: Use parenthesis on inside the comparison
```suggestion
    if (LINE_SENSOR_CHANNEL_DLC == (maxLineSensors * sizeof(uint16_t)))
```

File: lib/APPReinforcementLearning/src/App.cpp [hunk 45]
```diff
@@ -137,10 +188,89 @@ void App::loop()
...
+        while (maxLineSensors > lineSensorIdx)
+        {
+            payload.lineSensorData[lineSensorIdx] = lineSensorValues[lineSensorIdx];
+
+            ++lineSensorIdx;
+        }
+    }
+
+    /* Ignoring return value, as error handling is not available. */
+    (void)m_smpServer.sendData(m_serialMuxProtChannelIdLineSensors, &payload, sizeof(payload));
```
Comment (gabryelreyes) [inline lib/APPReinforcementLearning/src/App.cpp:210]: Are you sure you do not want to handle the errors? What happens if the messages are not sent?

File: lib/APPReinforcementLearning/src/App.cpp [hunk 46]
```diff
@@ -106,29 +110,76 @@ void App::loop()
...
+    {
+        uint8_t mode_options = ReadyState::getInstance().setSelectedMode();
+
+        if (mode_options > 0)
+        {
+            SMPChannelPayload::Mode payload =
+                (mode_options == 1) ? SMPChannelPayload::Mode::DRIVING_MODE : SMPChannelPayload::Mode::TRAINING_MODE;
+
+            /* Ignoring return value, as error handling is not available. */
+            (void)m_smpServer.sendData(m_serialMuxProtChannelIdMode, &payload, sizeof(payload));
```
Comment (gabryelreyes) [inline lib/APPReinforcementLearning/src/App.cpp:152]: Are you sure you do not want to handle the errors? What happens if the messages are not sent?

File: lib/APPReinforcementLearning/src/App.cpp (hunk 42 above)
Comment (gabryelreyes) [inline lib/APPReinforcementLearning/src/App.cpp:146]: Add postfix to the number
```suggestion
        if (0U < mode_options)
```

File: lib/APPReinforcementLearning/src/ReadyState.h [hunk 47]
```diff
@@ -96,19 +97,67 @@ class ReadyState : public IState
      */
     void setLapTime(uint32_t lapTime);
 
+    /**
+     *  Set the selected mode.
+     *
+     * @return It returns 1 if DrivingMode is selected or 2 if TrainingMode is selected.
+     */
+    uint8_t setSelectedMode();
```
Comment (gabryelreyes) [inline lib/APPReinforcementLearning/src/ReadyState.h:105]: This should be a `get` instead of a `set`, according to the implementation and the description

**Review by gabryelreyes — CHANGES_REQUESTED** [review]

**Review by gabryelreyes — APPROVED** [review]

**Review by hoeftjch — APPROVED** [review]
looks good to me now. Architecture changes in RL_Supervisor will likely be necessary for next updates.
